File: packages/server/internal/game/block/block.go
package block

type Num uint16
type Owner uint16
type Name string

type Meta struct {
	Name        Name
	Description string
}

type AllowMove struct {
	From   bool
	To     bool
	Reason string // debug info
}

type Block interface {
	Num() Num
	Owner() Owner

	// Round Events
	RoundStart(roundNum uint16)
	RoundEnd(roundNum uint16)

	// Move related
	AllowMove() AllowMove
	MoveFrom(num Num) Num
	// MoveTo Ret: a new block to replace this place
	MoveTo(num Num, owner Owner) Block

	Fog(isOwner bool, isSight bool) Block

	Meta() Meta
}

type Position struct{ X, Y uint8 }

File: packages/shared-types/src/block.ts
// Type aliases for numeric types
export type Num = number; // uint16 in Go
export type Owner = number; // uint16 in Go
export type Name = string;

// Meta interface
export interface Meta {
  name: Name;
  description: string;
}

// AllowMove interface
export interface AllowMove {
  from: boolean;
  to: boolean;
  reason: string; // debug info
}

// Block interface
export interface Block {
  num(): Num;
  owner(): Owner;
  
  // Round Events
  roundStart(roundNum: number): void;
  roundEnd(roundNum: number): void;
  
  // Move related
  allowMove(): AllowMove;
  moveFrom(num: Num): Num;
  // MoveTo returns a new block to replace this place
  moveTo(num: Num, owner: Owner): Block;
  
  fog(isOwner: boolean, isSight: boolean): Block;
  
  meta(): Meta;
}
